v2/cmd/asoctl/internal/importing/resource_importer.go

```diff
@@ -76,82 +74,45 @@ func (ri *ResourceImporter) AddARMID(armID string) error {
-// This allows us to call it from other methods that already have the lock
-func (ri *ResourceImporter) addImpl(importer ImportableResource) {
-	// If we've already seen this resource, skip it
-	// This happens frequently with extension resources as they can be inherited onto many regular resources
-	// and we don't want to attempt to import them more than once.
-	// It can also happen with regular resources if they are referenced by multiple other resources.
-	if ri.unique.Contains(importer.Id()) {
-		return
-	}
-
-	// Add it to our map and our queue
-	ri.queue = append(ri.queue, importer.Id())
-	ri.pending[importer.Id()] = importer
-	ri.unique.Add(importer.Id())
-}
-
 // Import imports all the resources that have been added to the importer.
 // Partial results are returned even in the case of an error.
 func (ri *ResourceImporter) Import(
 	ctx context.Context,
 ) (*ResourceImportResult, error) {
-	var processed int64 = 0
 
-	globalBar := ri.progress.AddBar(
-		0, // zero total because we don't know how much work will be needed,
-		mpb.PrependDecorators(
-			decor.Name("Import Azure Resources", decor.WCSyncSpaceR)),
-		mpb.AppendDecorators(decor.Percentage(decor.WC{W: 5})))
-	globalBar.SetPriority(-1) // must display at the top
-
-	// Why the nested loop?
-	// We can potentially trigger all the pending resources listed in ri.queue to be imported in parallel
-	// and end up with nothing on the queue because they're still running.
-	// The outer loop gives us a last look at the queue and allows us to keep going if there are new items
-	// requiring import.
-	var errs []error
-	for len(ri.queue) > 0 {
-		var eg errgroup.Group
-
-		// Limit concurrent processing, so we don't do too much at once
-		// The exact value isn't too important
-		eg.SetLimit(8)
-
-		for {
-			if ctx.Err() != nil {
-				// Context cancelled, time to stop
-				return nil, errors.Wrapf(ctx.Err(), "cancelled while importing")
-			}
+	workers := 1
+	candidates := make(chan ImportableResource)  // candidates that need to be deduped
+	pending := make(chan ImportableResource)     // importers that are pending import
+	completed := make(chan ImportResourceResult) // importers that have been executed successfully
+	done := make(chan struct{})                  // signal that we're done
 
-			rsrc, ok := ri.DequeueResource()
-			if !ok {
-				// No more resources to import
-				break
-			}
+	progress := ri.createProgressBar("Import Azure Resources", done)
 
-			eg.Go(func() error {
-				err := ri.ImportResource(ctx, rsrc)
+	// Dedupe candidates so we import each distinct resource only once
+	go ri.queueUniqueImporters(candidates, pending, progress)
 
-				// Update our global progress too
-				processed++
-				globalBar.SetTotal(processed+int64(len(ri.queue)), false)
-				globalBar.SetCurrent(processed)
+	// Create workers to run the import
+	for i := 0; i < workers; i++ {
+		go ri.importWorker(ctx, pending, completed, progress)
```

Comment:
It's been a while since I've worked with channels and go concurrency.

If I'm not wrong, having multiple sources pushing into a channel at the same time and not having enough capacity to read from a channel would result in data loss. Would suggest using buffered channel for progress in case of concurrent writes.

Reply:
@babbageclunk convinced me that unbuffered channels are preferable, partly because they make the code more deterministic. Adding a buffer can mask (but won't eliminate) race conditions. There's no data loss with this approach.